crates/payout_curve/src/lib.rs

```diff
@@ -0,0 +1,685 @@
+) -> Result<Vec<(PayoutPoint, PayoutPoint)>> {
+    let mut pieces = vec![];
+    let total_collateral = coordinator_collateral + trader_collateral;
+
+    // 1. we need to calculate the lower and upper bounds. This depends on the direction of the
+    let (liquidation_price_lower_bound, liquidation_price_upper_bound) =
+        calculate_payout_curve_bounds(
+            initial_price,
+            leverage_trader,
+            leverage_coordinator,
+            coordinator_direction,
+        )?;
+
+    let (short_leverage, long_leverage) = {
+        if coordinator_direction == Direction::Long {
+            (leverage_trader, leverage_coordinator)
+        } else {
+            (leverage_coordinator, leverage_trader)
+        }
+    };
+
+    let (lower_range_lower, lower_range_upper) = calculate_lower_range_payouts(
+        coordinator_direction,
+        total_collateral,
+        liquidation_price_lower_bound,
+    )?;
+
+    let last_payout = lower_range_upper.clone();
+
+    pieces.push((lower_range_lower, lower_range_upper));
+
+    let mid_range = calculate_mid_range_payouts(
+        trader_collateral,
+        coordinator_collateral,
+        long_leverage,
+        short_leverage,
+        initial_price,
+        liquidation_price_lower_bound
+            .to_u64()
+            .expect("to fit dec into u64"),
+        liquidation_price_upper_bound
+            .to_u64()
+            .expect("to fit dec into u64"),
+        &last_payout,
+        coordinator_direction,
+        quantity,
+    )?;
+
+    let (_, last_upper_range) = mid_range
+        .last()
+        .context("expect to have at least two elements in payout curve")?
+        .clone();
+
+    for (lower, upper) in mid_range {
+        pieces.push((lower, upper));
+    }
+    // if the upper bound is already [`BTCUSD_MAX_PRICE`] we don't have to add the upper bound
+    // anymore
+    if last_upper_range.event_outcome < BTCUSD_MAX_PRICE {
+        let upper_range_payout_points = calculate_upper_range_payouts(
+            coordinator_direction,
+            total_collateral,
+            last_upper_range,
+        )?;
+        pieces.push(upper_range_payout_points);
+    }
+
+    Ok(pieces)
+}
+
+/// Calculates the mid range payout points between lower liquidation and upper liquidation
+///
+/// Returns tuples of payout points, first item is lower point, next item is higher point of two
+/// points on the payout curve
+#[allow(clippy::too_many_arguments)]
+fn calculate_mid_range_payouts(
+    trader_collateral: u64,
+    coordinator_collateral: u64,
+    long_leverage: f32,
+    short_leverage: f32,
+    initial_price: Decimal,
+    lower_limit: u64,
+    upper_limit: u64,
+    last_payout: &PayoutPoint,
+    direction: Direction,
+    quantity: f32,
+) -> Result<Vec<(PayoutPoint, PayoutPoint)>> {
+    let total_collateral = trader_collateral + coordinator_collateral;
+
+    let pieces = (lower_limit..=upper_limit - 1)
+        .step_by(2)
```

Comment:
❓ Is this the reason why the curve generation is so slow? We probably don't need such a small step when we are discretising.

Reply:
I assume so, yes